### PR TITLE
optional offset argument to WebSocket streams

### LINKED ISSUE #4509: bookmarks in query streams, with offset argument to websocket endpoints
Starting at the end of non-live data (#4461) we will include a `"offset"` as discussed in #3878. This will incompatibly change the liveness marker from `{"live": true}` to the first occurrence of an offset: `{"events": [...], "offset": ...}`, where `[...]` *may* be empty.

As implied by the above, not every response block will include an offset. We only promise to provide offsets once we reach the end of non-live data. In fact, we promise *not* to provide offsets until we reach the end of non-live data.

The current query union format is `{templateIds, query} | [{templateIds, query}...]`, and fetch by key format is `[{templateId, key}]`; @hurryabit to determine where the nicest place to put `offset` is in these datatypes, noting that we can only reasonably have one offset for all queries in the union. The behavior of specifying a non-existent offset will be passed through from ledger API.

The only intended effect of this change is to make streams reconnect-able; the remaining stream semantics will not change.

This will disable "phantom archive" filtering entirely for resumed contract key streams, while it will remain in place for fetch streams for which offset is not supplied. Refining this behavior should be addressed in a separate issue, as @gerolf-da has noted various complicating factors.

Depends on #4465.

cc @dtanabe @lima-da 

* [ ] how to pass offset, especially in case of `[union...]`?
* [x] how to represent `LedgerBegin`? `null`

### PR DESCRIPTION
**Pull request by S11001001**
Fixes #4509.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/json-api/index.rst [hunk 1]
```diff
@@ -1021,6 +1021,13 @@ Contracts Query Stream
 List currently active contracts that match a given query, with
 continuous updates.
 
+An optional ``offset`` returned by a prior query (see output examples
+below) may be specified *first*, as a separate body.  It must be a
+string, and if specified, the stream will begin immediately *after* the
+response body that included that offset::
+
+    {"offset": "5609"}
```
Comment (hurryabit) [inline docs/source/json-api/index.rst]: I would move that block after the description of the "main" query format and replace *first* by something *before* the "main" query. My rationale is that I want the simple use case to come first and the more advanced stuff later. Other than that, the docs and the API are fine. Thanks a lot.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala [hunk 2]
```diff
@@ -272,27 +272,36 @@ class ContractsService(
...
       getCreatesAndArchivesSince(jwt, txnFilter, _: api.ledger_offset.LedgerOffset, terminates)
 
-    import ContractsFetch.acsFollowingAndBoundary, ContractsFetch.GraphExtensions._
-    val contractsAndBoundary = acsFollowingAndBoundary(transactionsSince).divertToHead
-    source
-      .viaMat(contractsAndBoundary) { (nu, fob) =>
-        fob.foreach(a => logger.debug(s"contracts fetch completed at: ${a.toString}"))
-        nu
-      }
+    // TODO SC use startOffset
```
Comment (leo-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala]: why TODO? you are already sourcing the `startOffset` into `transactionsFollowingBoundary` flow.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala [hunk 3]
```diff
@@ -290,28 +317,38 @@ class WebSocketService(
...
           Future successful -\/(
             InvalidUserInput("Cannot process your input, Expect a single JSON message"))
       }
+      .via(withOptPrefix { ejv: (InvalidUserInput \/ JsValue) =>
+        ejv
+          .flatMap(jv =>
+            Bitraverse[\/].bisequence(
+              readStartingOffset(jv) toLeftDisjunction Q.parse(decoder, jv).strengthL(None)))
+          .sequence: domain.StartingOffset \/ (Error \/ (Option[domain.StartingOffset], A))
+      }((offPrefix, ejv) => ejv flatMap (jv => Q.parse(decoder, jv) strengthL Some(offPrefix))))
```
Comment (leo-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala]: this entire block + `readStartingOffset` could have been simplified with derived JSON Protofol for `domain.StreamingRequest` variant. Hard to comprehend what is going on here.

Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala]: You might be right about `readStartingOffset`, but that would do nothing to simplify this block.

Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala]: The reason I didn't derive is that I do not want to treat all JSON parse errors equally;

1. a missing ``offset`` key should mean skip to the required body,
2. a present ``offset`` key with non-string should fail, and
3. mixing an ``offset`` key with any other key should fail,

but when you derive parsers

```
final case class Foo(foo: Int)
"""{"foo":1, "offset":"hi"}""".parseJson.convertTo[Foo]
// ⇑ evaluates to Foo(1)
```

Once you add all this special handling, bothering with the derived parser becomes less a convenience and more a...bother. I can still work it in if you'd prefer, but it's not going to be as pretty as you might be imagining. Let me know what you think.

I'll also think about what I can do with the definition of `withOptPrefix`, which I believe can be made more complicated to simplify the call to it that you've directly highlighted.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala (hunk 2 above)
Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala]: Just forgot to delete it :)

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala (hunk 3 above)
Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala]: See 9abd4e8cd8 for a `withOptPrefix` redesign.

File: docs/source/json-api/index.rst (hunk 1 above)
Comment (S11001001) [reply docs/source/json-api/index.rst]: Moved to just before we start talking about output in e6995900820f8938086428468eb8627902072729

**Pull request merged**